src/cdomains/intDomain.ml

```diff
@@ -1862,32 +1862,49 @@ module Booleans = MakeBooleans (
   end)
 
 (* Inclusion/Exclusion sets. Go to top on arithmetic operations after ana.int.enums_max values. Joins on widen, i.e. precise integers as long as not derived from arithmetic expressions. *)
-module Enums : IkindUnawareS = struct
+module Enums : S with type int_t = BigInt.t = struct
   open Batteries
-  module I = Integers
+  module I = BigInt
   module R = Interval32 (* range for exclusion *)
 
   let range_ikind = Cil.IInt
   let size t = R.of_interval range_ikind (let a,b = Size.bits_i64 t in Int64.neg a,b)
   type e = I.t (* element *)
-  and t = Inc of e list | Exc of e list * R.t [@@deriving to_yojson] (* inclusion/exclusion set *)
-  type int_t = int64
+  and t = Inc of I.t list | Exc of I.t list * R.t [@@deriving to_yojson] (* inclusion/exclusion set *)
```

Comment:
I feel like we should be using actual sets here, not lists. It's probably annoying to rewrite though because singleton `Inc`s cannot be then pattern matched.

Reply:
With 471f833356b253f7f5730f1c6cc8f8855d3bf0b3 the domain is now using sets, but this caused a termination problem of the solver in `01/39`. I am not sure what caused this. Termination on the tests work when not doing narrowing for Enums (see  96bb4a08d07f38bcb9be83ab0fa9e02018f59bc3).

Reply:
It would be very interesting to see where this happens, it is a bit surprising given that just switching from representing things as lists to representing them as sets should not cause any changes in termination. :thinking: 

I'd somehow expect that issue to be tied to the loop in the program
https://github.com/goblint/analyzer/blob/20f3a7d6285a35c7c7d5c2f6206e280a73b83c25/tests/regression/01-cpa/39-interval-and-enums.c#L52-L58

Reply:
Yes, the solver is looping when analyzing this `for`-loop. 

The non-termation was introduced by the changes to [Enums here](https://github.com/goblint/analyzer/pull/175/commits/471f833356b253f7f5730f1c6cc8f8855d3bf0b3#diff-7c2aff56950cfd3ab8421eed789e8ee63632f69397ffbe53db74ae4729641fa3); I don't see what could possibly cause this.

Reply:
I think I found the problem. The issue is caused because the `hash` and `compare` for `Enums` now have to be implemented differently; probably because structural equality does not work for `Set`.

Reply:
Yes, we used sorted lists because it should be cheaper than sets which have no cheap `equal`/`compare` (O(1) vs. O(n log n)?).
And because it allows for easier pattern matching.
Did you check if it hashconsing still works for equal sets with different repr.?
What speaks for using sets?

Reply:
`DefExc` already uses sets, so if there's any problem with that, it existed already there. So if we want to optimize, we'd probably want to optimize both simultaneously. It wouldn't be very difficult to give a sorted list based `SmallSet` implementation of the same signature.

Reply:
Yes, probably makes sense. At the time I implemented Enums I didn't want to change DefExc.
Maybe first check these assumptions:
- physical and structural equality for lists are both O(1)
- set equality is done by fold mem which is O(n * log n)
- we call `equal`/`compare` way more often than we insert elements (which has the same complexity for insertion sorted lists)
- we have cases where we get sets that are equal but not structurally equal - how? Just inserting elements in different order shouldn't be enough since it's binary trees - combinations of insert/remove?